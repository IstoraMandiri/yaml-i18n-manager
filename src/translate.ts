import googleTranslate from 'google-translate';

import type { ExistingYaml } from './types';

if (!process.env.GT_KEY) {
  throw new Error('You must specify a google translate key `export GT_KEY=XXX`');
}

const gt = googleTranslate(process.env.GT_KEY);

function transformTranslation(translated: string, original: string) {
  let str = translated
    // URL formatting
    .replace(/] \(/g, '](')
    .replace(/\/ /g, '/')
    .replace(/ \//g, '/')
    .replace(/! \[/g, '![')
    // chinese characters
    .replace(/＃/g, '#')
    .replace(/！/g, '!')
    .replace(/（/g, '(')
    .replace(/）/g, ')')
    // markdown multiline titles (##X => ## X)
    .replace(/^#+(?=[^\s#])/gm, '$& ');

  // line breaks
  if (original.endsWith('  ')) {
    str = `${str}  `;
  }

  return str;
}

export async function translateYaml(
  normalized: ExistingYaml[],
  from: string,
  to: string,
): Promise<any> {
  console.log(`Using google to translate ${normalized.length} items`);
  return await new Promise((resolve, reject) => {
    gt.translate(
      normalized.map(({ value }) => value),
      from,
      to,
      (err: Error, _res: { translatedText: string }[]) => {
        if (err) {
          console.log(err);
          return reject(err);
        }
        const res = Array.isArray(_res) ? _res : [_res];
        return resolve(
          res.map(({ translatedText: translated }, i) => {
            const { vHash, value: defaultValue, ...original } = normalized[i];
            const value = transformTranslation(translated, defaultValue);
            return { ...original, dHash: vHash, defaultValue, value, locale: to };
          }),
        );
      },
    );
  });
}
