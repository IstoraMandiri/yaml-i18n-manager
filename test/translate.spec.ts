jest.mock('google-translate');
import googleTranslate from 'google-translate';

googleTranslate.mockImplementation(() => {
  return {
    translate: (
      strings: string[],
      _from: string,
      _to: string,
      callback: (err: null, res: { translatedText: string }[]) => {},
    ) => {
      callback(
        null,
        strings.map((s) => ({ translatedText: `${s} ] (/  /! [＃！（）` })),
      );
    },
  };
});

import translate from '../src/translate';

const sampleData = [
  {
    path: 'string',
    locale: 'string',
    sig: 'string',
    key: 'string',
    value: 'string',
    vHash: 'string',
  },
  {
    path: 'string2',
    locale: 'string2',
    sig: 'string2',
    key: 'string2',
    value: 'string2',
    vHash: 'string2',
  },
];

describe('translate', () => {
  it('returns the correct translated YAML object', async () => {
    expect(await translate(sampleData, 'en', 'jp')).toMatchSnapshot();
  });
});
