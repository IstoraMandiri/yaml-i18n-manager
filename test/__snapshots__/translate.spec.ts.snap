// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translate returns the correct translated YAML object 1`] = `
Array [
  Object {
    "dHash": "string",
    "defaultValue": "string",
    "key": "string",
    "locale": "jp",
    "path": "string",
    "sig": "string",
    "value": "string ](//![#!()",
  },
  Object {
    "dHash": "string2",
    "defaultValue": "string2",
    "key": "string2",
    "locale": "jp",
    "path": "string2",
    "sig": "string2",
    "value": "string2 ](//![#!()",
  },
]
`;
